### PR TITLE
Laboratory work #4, Ogabek Matyoqubov - 22FPL2

### PR DESCRIPTION
**Pull request by ogabekmatyoqubov**

### CODE REVIEW

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]
What about start.py?

File: lab_4_fill_words_by_ngrams/main.py [hunk 1]
```diff
@@ -28,6 +31,22 @@ def _tokenize(self, text: str) -> tuple[str, ...]:  # type: ignore
...
             ValueError: In case of inappropriate type input argument or if input argument is empty.
         """
+        if not isinstance(text, str) or not text:
+            raise ValueError
+        for symbol in ['.', '!', '?']:
+            text = text.replace(symbol, ' ' + self._end_of_word_token)
+        tokens = []
+        for word in text.lower().split():
+            if word == self._end_of_word_token or word.isalpha():
+                tokens.append(word)
```
Comment (artyomtugaryov) [inline lab_4_fill_words_by_ngrams/main.py:41]: use `continue` and remove `else`

File: lab_4_fill_words_by_ngrams/main.py [hunk 2]
```diff
@@ -28,6 +31,22 @@ def _tokenize(self, text: str) -> tuple[str, ...]:  # type: ignore
         Raises:
             ValueError: In case of inappropriate type input argument or if input argument is empty.
         """
+        if not isinstance(text, str) or not text:
+            raise ValueError
```
Comment (artyomtugaryov) [inline lab_4_fill_words_by_ngrams/main.py:35]: add message

File: lab_4_fill_words_by_ngrams/main.py [hunk 3]
```diff
@@ -64,7 +93,9 @@ class TopPGenerator:
 
     Attributes:
         _model (NGramLanguageModel): NGramLanguageModel instance to use for text generation
+    generation
```
Comment (artyomtugaryov) [inline lab_4_fill_words_by_ngrams/main.py:96]: ?

File: lab_4_fill_words_by_ngrams/main.py [hunk 4]
```diff
@@ -64,7 +93,9 @@ class TopPGenerator:
...
         _word_processor (WordProcessor): WordProcessor instance to handle text processing
+    processing
```
Comment (artyomtugaryov) [inline lab_4_fill_words_by_ngrams/main.py:98]: ?

File: lab_4_fill_words_by_ngrams/main.py [hunk 5]
```diff
@@ -125,6 +187,8 @@ def get_conversion_generator_type(self, generator_type: int) -> str:  # type: ig
         Returns:
             (str): Name of the generator.
         """
+        generators = ['Greedy Generator', 'Top-P Generator', 'Beam Search Generator']
+        return generators[generator_type]
```
Comment (artyomtugaryov) [inline lab_4_fill_words_by_ngrams/main.py:191]: what if `generation_type` > 3?

File: lab_4_fill_words_by_ngrams/main.py [hunk 6]
```diff
@@ -155,14 +222,15 @@ def get_perplexity(self) -> float:  # type: ignore
         Returns:
             (float): Perplexity value
         """
-
+        return self.__perplexity
```
Comment (artyomtugaryov) [inline lab_4_fill_words_by_ngrams/main.py:225]: add new line

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]
Add message to each error

**Comment by vitazaytseva** [comment]
Well done! Good job!

**Pull request closed without merging**